File: frontend/src/types/risk.ts
import type { PermitType, PriorityLevel, RecommendationRead, RiskEventRead, SeverityLevel } from "@/types/domain";

export interface SensorSignalInput {
  sensor_id?: string | null;
  sensor_name?: string | null;
  sensor_type?: string | null;
  value: number;
  unit?: string | null;
  timestamp?: string | null;
}

export interface RuleConditionResult {
  field: string;
  operator: string;
  expected: unknown;
  observed: unknown;
  matched: boolean;
  label?: string | null;
}

export interface RuleMatch {
  rule_id: string;
  name: string;
  description: string;
  severity: SeverityLevel;
  risk_category: string;
  score_delta: number;
  reason: string;
  regulations: string[];
  conditions: RuleConditionResult[];
}

export interface RecommendationAction {
  action: string;
  priority: PriorityLevel;
  rationale: string;
  target_type?: string | null;
  target_id?: string | null;
  source_rules: string[];
}

export interface HistoricalSimilarityMatch {
  source: string;
  reference_id: string;
  title: string;
  similarity_score: number;
  summary: string;
  evidence: string[];
}

export interface ExplainabilityReport {
  why: string;
  evidence: Array<Record<string, unknown>>;
  contributing_factors: string[];
  applicable_rules: RuleMatch[];
  historical_similarity: HistoricalSimilarityMatch[];
  recommended_actions: RecommendationAction[];
  impact_summary?: Record<string, unknown> | null;
}

export interface RiskAnalysisRequest {
  zone_id?: string | null;
  equipment_id?: string | null;
  gas_level?: number | null;
  temperature?: number | null;
  pressure?: number | null;
  humidity?: number | null;
  vibration?: number | null;
  equipment_health?: number | null;
  permit_type?: PermitType | null;
  worker_count?: number;
  worker_present?: boolean;
  worker_ids?: string[];
  maintenance_running?: boolean;
  maintenance_overdue?: boolean;
  weather_condition?: string | null;
  weather_temperature_c?: number | null;
  weather_humidity?: number | null;
  wind_kph?: number | null;
  historical_similarity?: number | null;
  shift?: string | null;
  time_of_day?: string | null;
  sensor_readings?: SensorSignalInput[];
  persist_result?: boolean;
}

export interface RiskAnalysisResult {
  risk_score: number;
  severity: SeverityLevel;
  confidence: number;
  risk_category: string;
  reason: string;
  root_cause: string;
  recommendation: string;
  expected_consequence: string;
  regulations: string[];
  evidence: Array<Record<string, unknown>>;
  applicable_rules: RuleMatch[];
  historical_similarity: HistoricalSimilarityMatch[];
  recommended_actions: RecommendationAction[];
  explainability: ExplainabilityReport;
  graph_insights?: Record<string, unknown> | null;
  recommendations_created: RecommendationRead[];
  risk_event?: RiskEventRead | null;
}

File: backend/app/schemas/risk.py
from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field

from app.models.enums import PermitType, PriorityLevel, SeverityLevel
from app.schemas.domain import RecommendationRead, RiskEventRead


class SensorSignalInput(BaseModel):
    sensor_id: str | None = None
    sensor_name: str | None = None
    sensor_type: str | None = None
    value: float
    unit: str | None = None
    timestamp: datetime | None = None


class RuleConditionResult(BaseModel):
    field: str
    operator: str
    expected: Any
    observed: Any
    matched: bool
    label: str | None = None


class RuleMatch(BaseModel):
    rule_id: str
    name: str
    description: str
    severity: SeverityLevel
    risk_category: str
    score_delta: float
    reason: str
    regulations: list[str] = Field(default_factory=list)
    conditions: list[RuleConditionResult] = Field(default_factory=list)


class RecommendationAction(BaseModel):
    action: str
    priority: PriorityLevel
    rationale: str
    target_type: str | None = None
    target_id: str | None = None
    source_rules: list[str] = Field(default_factory=list)


class HistoricalSimilarityMatch(BaseModel):
    source: str
    reference_id: str
    title: str
    similarity_score: float
    summary: str
    evidence: list[str] = Field(default_factory=list)


class ExplainabilityReport(BaseModel):
    why: str
    evidence: list[dict[str, Any]] = Field(default_factory=list)
    contributing_factors: list[str] = Field(default_factory=list)
    applicable_rules: list[RuleMatch] = Field(default_factory=list)
    historical_similarity: list[HistoricalSimilarityMatch] = Field(default_factory=list)
    recommended_actions: list[RecommendationAction] = Field(default_factory=list)
    impact_summary: dict[str, Any] | None = None


class RiskAnalysisRequest(BaseModel):
    zone_id: str | None = None
    equipment_id: str | None = None
    gas_level: float | None = None
    temperature: float | None = None
    pressure: float | None = None
    humidity: float | None = None
    vibration: float | None = None
    equipment_health: float | None = None
    permit_type: PermitType | None = None
    worker_count: int = 0
    worker_present: bool = False
    worker_ids: list[str] = Field(default_factory=list)
    maintenance_running: bool = False
    maintenance_overdue: bool = False
    weather_condition: str | None = None
    weather_temperature_c: float | None = None
    weather_humidity: float | None = None
    wind_kph: float | None = None
    historical_similarity: float | None = None
    shift: str | None = None
    time_of_day: str | None = None
    sensor_readings: list[SensorSignalInput] = Field(default_factory=list)
    persist_result: bool = True


class RiskAnalysisResult(BaseModel):
    risk_score: float
    severity: SeverityLevel
    confidence: float
    risk_category: str
    reason: str
    root_cause: str
    recommendation: str
    expected_consequence: str
    regulations: list[str] = Field(default_factory=list)
    evidence: list[dict[str, Any]] = Field(default_factory=list)
    applicable_rules: list[RuleMatch] = Field(default_factory=list)
    historical_similarity: list[HistoricalSimilarityMatch] = Field(default_factory=list)
    recommended_actions: list[RecommendationAction] = Field(default_factory=list)
    explainability: ExplainabilityReport
    graph_insights: dict[str, Any] | None = None
    recommendations_created: list[RecommendationRead] = Field(default_factory=list)
    risk_event: RiskEventRead | None = None
